bip-ccv.mediawiki

```diff
@@ -0,0 +1,420 @@
+
+-----
+
+Note that the ''deduct'' semantic does not allow to check the exact amount of its output. Therefore, in contracts using a scheme similar to figure 3 or 4 above, amounts be constrained with a signature, or with future introspection opcodes that allow fixing the amount.
+
+== Specification ==
+
+The tapscript opcode <code>OP_SUCCESS187</code> (<code>0xbb</code>) is constrained with new rules to implement <code>OP_CHECKCONTRACTVERIFY</code>.
+
+When evaluating <code>OP_CHECKCONTRACTVERIFY</code> (<code>OP_SUCCESS187</code>,
+<code>0xbb</code>), the expected format of the stack, shown top to bottom, is:
+
+<source>
+<flags>
+<taptree>
+<pk>
+<index>
+<data>
+</source>
+
+where:
+
+* <code><flags></code> is a minimally encoded integer, according to one of the values defined below.
+* <code><taptree></code> is the Merkle root of the taproot tree, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><pk></code> is called the ''naked key'', and it's a valid 32-byte x-only public key, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><index></code> is a minimally encoded -1, or a minimally encoded non-negative integer.
+* <code><data></code> is a buffer of arbitrary length.
+
+In short, the semantics of the opcode with respect to the Script can be summarized as follows:
+
+<blockquote>
+Verify that the input/output with the given 'index' is a P2TR UTXO whose taproot output key is obtained from 'pk', tweaked with the hash of 'data' (if non-empty), then taptweaked with 'taptree' (if non-empty).
+</blockquote>
+
+If the <code><data></code> is non-empty, then the additive tweak for the data is computed as:
+
+<source lang="python">
+  data_tweak = sha256(pk || data)
+</source>
+
+In the following, the ''current input'' is the input whose Script is being executed.
+
+The following value of the <code><flags></code> are defined:
+* <code>CCV_FLAG_CHECK_INPUT = -1</code>: Check an input's script; no amount check.
+* <code>CCV_FLAG_CHECK_OUTPUT = 0</code>: Check an output's script; preserve the (possibly residual) amount.
+* <code>CCV_FLAG_IGNORE_OUTPUT_AMOUNT = 1</code>: Check an output's script; ignore amounts.
+* <code>CCV_FLAG_DEDUCT_OUTPUT_AMOUNT = 2</code>: Check an output's script; deduct the output amount from the input's residual amount.
+
+Any other value of the <code><flags></code> makes the opcode succeed validation immediately for the current input<ref>This allows to soft-fork future behavior by introducing new values for the <code><flags></code>. As the flags would always be hard-coded via a push in the Script, the risk of mistakes seems negligible.</ref>.
+
+The following values of the other parameters have special meanings:
+* If the <code><taptree></code> is -1, it is replaced with the Merkle root of the current input's tapscript tree. If the taptree is the empty buffer, then the taptweak is skipped.
```

Comment:
Have you considered using flags for these special cases and then having fewer stack elements? That seems both more space efficient and probably makes scripts easier to read.

Reply:
All the flags and the special parameters are encoded as 1-byte push opcodes, so I don't think you can really save much by aggregating them in the flags (e.g. using extra bits that are currently in the `OP_SUCCESS` behavior).
Moreover, I think it would be very inconvenient (both in implementation and in Script readability) if the number of stack arguments is different for different use cases of CCV.